server/apps/immich/src/app.module.ts

```diff
@@ -15,20 +14,23 @@ import { AlbumModule } from './api-v1/album/album.module';
 import { AppController } from './app.controller';
 import { ScheduleModule } from '@nestjs/schedule';
 import { ScheduleTasksModule } from './modules/schedule-tasks/schedule-tasks.module';
-import { DatabaseModule } from '@app/database';
 import { JobModule } from './api-v1/job/job.module';
 import { SystemConfigModule } from './api-v1/system-config/system-config.module';
 import { OAuthModule } from './api-v1/oauth/oauth.module';
 import { TagModule } from './api-v1/tag/tag.module';
 import { ImmichConfigModule } from '@app/immich-config';
 import { ShareModule } from './api-v1/share/share.module';
+import { DomainModule } from '@app/domain';
+import { InfraModule } from '@app/infra';
+import { UserController } from './controllers';
 
 @Module({
   imports: [
     ConfigModule.forRoot(immichAppConfig),
 
-    DatabaseModule,
-    UserModule,
+    DomainModule.register({
+      imports: [InfraModule],
```

Comment:
What is the difference between this and import the `InfraModule` into `AppModule`?

If the `InfraModule` is always needed in `DomainModule`. Why don't we have it import in the `DomainModule` by default?

Reply:
The domain module does have a dependency on the repository interface, but not that specific implementation.

We could move the dependency to the domain itself, but it is better for the domain to be self contained. Registering it this way means the domain does not have any imports from any other modules.

In the future this should also allow us to do black box testing of the domain without needing an actual database.

Reply:
Another way of looking at it: Domain doesn't depend specifically on the infra module, just the providers is exports and we may make another module for testing that provides the same things. Using a dynamic module means both situations are feasible without changing anything in the domain module. It doesn't really care where the providers come from as long as they're provided.

Reply:
We actually already test the domain without needing the actual database thanks to the hexagonalish architecture (we test from the services).

I don't think we will ever inject a different infra module, so maybe is not such a bad idea to inject by default the infra module. The only dependency with the `@app/infra` is at the "module" layer which is actually in the framework side of thinks, so outside the domain.

I'm happy with any approach though, just pointing out that I think we will be always injecting the same infra module, and it doesn't have any benefit for testing 😃

Reply:
Yeah that's true. Maybe there isn't a scenario in which we'd mock all the providers at once. 

I can try making it a direct dependency. 

The only other potential issue would be circular dependencies between the two. Infra depends on the interface and the domain depends on the implementations.

Reply:
I am OK to leave it as is. I was just curious. My question is answered so we don't have to change it.